vcenter/vcenter.py

```diff
@@ -449,6 +442,120 @@ def stop(self, uuid):
+                                                                    custom_local_path=temp_screenshot)
+                if screenshot_path is not None:
+                    with open(screenshot_path, 'rb') as s:
+                        screenshot_content = base64.b64encode(s.read())
+
+            return screenshot_content
+
+    # TODO rewrite others to use this one
+    def __get_objects_list_from_container(self, container, object_type):
+
+        result = []
+        object_view = None
+        try:
+            object_view = self.content.viewManager.CreateContainerView(
+                    container,
+                    [object_type],
+                    True)
+            result = list(object_view.view)
+        except vmodl.fault.ManagedObjectNotFound:
+            self.__logger.warn('vmodl.fault.ManagedObjectNotFound has occured')
+        except Exception:
+            settings.raven.captureException(exc_info=True)
+        finally:
+            if object_view is not None:
+                object_view.Destroy()
+
+        return result
+
+    def __get_datacenter_for_datastore(self, datastore_name):
+        datacenters = self.__get_objects_list_from_container(self.content.rootFolder, vim.Datacenter)
+        for dc in datacenters:
+            datastores = self.__get_objects_list_from_container(dc, vim.Datastore)
+            for ds in datastores:
+                if ds.info.name == datastore_name:
+                    return dc
+        return None
+
+    def download_file_from_datastore(self, remote_path_to_file, datastore_name, custom_local_path=None):
+        """
+        Downloads file from datastore (with retries)
+        :param remote_path_to_file: path to file in datastore (e.g. my_vm/my_vm.png)
+        :param datastore_name: name of datastore
+        :param custom_local_path: Optional, is specified, file will be saved with this path.
+        Otherwise to current working directory using the same filename as on remote
+        :return: path to downloaded file, or None in case of failures
+        """
+        self.__check_connection()
+        server_name =  settings.app['vsphere']['host']
+        datacenter = self.__get_datacenter_for_datastore(datastore_name)
+        if datacenter is None:
+            raise RuntimeError(f'Cannot find datacenter for datastore {datastore_name}')
+
+        download_url = f'https://{server_name}/folder/{remote_path_to_file}?dcPath={datacenter.name}&dsName={datastore_name}'
+
+        for i in range(3):
+            try:
+                resp = requests.get(url=download_url, verify=False, headers={'Cookie': self._connection_cookie})
+                if resp.status_code == 200:
+                    # download ok, save return path
+                    downloaded_file_path = custom_local_path if custom_local_path is not None else os.path.basename(remote_path_to_file)
+                    with open(downloaded_file_path, 'wb') as lf:
+                        lf.write(resp.content)
```

Comment:
wouldn't it be better to return the data directly from the function instead of storing it to file and read it's content in caller function? I suppose that even when file size would be something like 2MB it is not problem to have it in a memory for a while.

Reply:
I like it better this way, but I can adjust it, if you wish ;)